frame/support/procedural/Cargo.toml

```diff
@@ -23,4 +23,5 @@ syn = { version = "1.0.58", features = ["full"] }
 
 [features]
 default = ["std"]
-std = []
+std = ["metadata-docs"]
+metadata-docs = []
```

Comment:
I wonder if it wouldn't be better to do `metadata-docs` in the scale-info crate
Instead of having to not insert docs, the scale-info crate would simply ignore all doc when the feature docs is not set.
Like we do for parity-scale-codec error "chain-error".

So in case `docs` is not set the types `Field` etc.. will not have `docs` fields, and all builder `fn docs` will do no-op, and maybe inlined to be sure the compiled code doesn't have the static string in the code.

(that said for parity-scale-codec I didn't verify that the functions string arguments were optimized away.)
But we can do in follow-up too

Reply:
`scale-info` does have the `docs` feature (although not implemented in the way you suggest, I might try that). We still need this feature here because the legacy macros are instantiating some metadata docs directly e.g.  in proc-macros via https://github.com/paritytech/substrate/blob/b61db7f052ceeffa4d55c1b77c0f5e048270ca67/frame/support/procedural/tools/src/docs.rs#L38 and legacy macros e.g. https://github.com/paritytech/substrate/blob/5de104ed5743307c2ee56f901f0cb972c17905d8/frame/support/src/dispatch.rs#L2476

So we could, as you suggest use a builder pattern for the docs on `frame-metadata` types and move the feature into that crate.

Reply:
yes, my idea would be that we always call `.doc(..)` to generate the type info. But frame-metadata would ignore those when built without doc.
I'm not against current implementation, but I feel like requires us to be more careful, for example the manual implementation in pallet identity of TypeInfo for IdentityField, doesn't build without doc and seems to always add the docs.
This is pretty minor but I think we can do better in a follow up.

Reply:
Yes this could be much better, I will fix it up in a follow up

Reply:
https://github.com/paritytech/scale-info/issues/109
https://github.com/paritytech/frame-metadata/issues/6